Implement 'rewrite' command to exclude files from existing snapshots

dionorgua opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

What is the purpose of this change? What does it change?
--------------------------------------------------------

This adds 'rewrite' command that provides a way to remove data from existing snapshots.

This is some sort of preliminary pull requests. It's actually works here on test repo, but lacks tests, etc. Mostly to discuss things

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

[ replaces #2720 because it was created from master branch ]
closes #14 

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [ ] I'm done, this Pull Request is ready for review

dionorgua committed:
Implement 'rewrite' command to exclude files from existing snapshots

aawsome commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -0,0 +1,296 @@
+	err = repo.SaveIndex(ctx)
+	if err != nil {
+		return false, err
+	}
```
The index is now automatically saved by `repo.Flush()` #2773, no need to call `SaveIndex` after `Flush` any more.

aawsome commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -0,0 +1,296 @@
+	var saveTreeFunc saveTreeFunction
+	if !opts.DryRun {
+		saveTreeFunc = func(tree *restic.Tree) (restic.ID, error) {
+			return repo.SaveTree(ctx, tree)
```
You could also use the TreeSaver from `internal/archiver` to make saving parallel. Can be however postponed to a later optimization.

MichaelEischer replied:
That would also require the BlobSaver. It's probably better to integrate some form of asynchronous blob submission directly into the repository and then use that.

aawsome commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -0,0 +1,296 @@
+		return true, nil
+	}
+
+	err = repo.Flush(ctx)
```
You only need to flush the repo when the snapshot was modified.

I also wouldn't flush after finishing a single snapshot, but once after all snapshots are rewritten, i.e. in line 289.
The reason is that flush may finish small packs. If you modify a lot of snapshots generating only few new tree blobs, you may end up in having lots of new packs which each only contain few tree blobs.

However, the snapshot files should only be written after flush finishes with nil error. So also writing the snapshots should be postponed. I suggest returning the new snapshot object here and write all snapshots after `Flush` in `runRewrite`

aawsome commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -0,0 +1,296 @@
+		return true, err
+	}
+
+	if opts.Inplace {
```
I would suggest to remove the deletion code totally from `cmd_rewrite.go` and instead call `runForget` from `forget.go` with the snapshot-IDs to remove.

That way you will not have any critical code that removes anything here, but keep that "burden" in exclusive pieces of code (which is well-tested and can be optimized or maybe turned into 2-phase-compaction later more easily).

Also allows to add a `--prune` flag here and pass this directly to `runForget`; we should then also add the `addPruneFlags()` call once #2718 is merged.

MichaelEischer replied:
Chaining the commands is probably more complex than directly removing the snapshot. In particular the error handling, log output etc. would have to be coordinated between the commands.

aawsome commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -0,0 +1,296 @@
+
+	if !gopts.NoLock && !opts.DryRun {
+		Verbosef("create exclusive lock for repository\n")
+		lock, err := lockRepoExclusive(repo)
```
If you outsource the deletion of files, a non-exclusive lock would suffice here.

MichaelEischer replied:
Upgrading a lock in the middle of a command is a problem. There is no guarantee that this will work.

aawsome reviewed (commented):
Looks mostly good to me :+1: 
I propose to call the existing code from `cmd_forget` to actually remove snapshots. That way also a pruning could be directly started from this command which IMO improves the usage for end users.

dionorgua commented:
Thanks a lot for review. I'll try to address everything soon

edmwagner commented:
I Just mirrored my minio and used this branch to rewrite a few different scenarios.  looks good so far.  restoring files worked as expected and matched versions restored from original repo.  :)  The repo is several GB smaller though.



I would like to help move this feature forward.

dionorgua commented:
Hi,
I was busy with personal things. Will try to catch up this weekend and at least update/rebase it.

Komic commented:
Is supporting such a feature for a whole repository planned?

aawsome commented:
> Is supporting such a feature for a whole repository planned?

This PR already allows to give multiple snapshots or even multiple snapshot criteria. E.g. you can run it with all snapshots or give multiply `-H` options to cover all hosts.

We could add a possibility to specify `all` which resolves to all snapshots - but I would leave that to a future PR as it is pretty independent from this change.

Komic commented:
Good stuff, thank you. 👍

bertbesser commented:
What's the status of this PR? Any chance that it gets merge soon?

rawtaz commented:
@bertbesser This PR should be considered work in progress. No decision has been made about whether this should be merged or not, the PR is not yet marked as done by the author, and there's a bunch of unaddressed comments. Clearly it's not ready for merging, and the status is simply what you see here in the GitHub GUI where you posted a comment. That said, there's a lot of other stuff going on if you look in this repository, so consider this WIP. Feel free to try it yourself though, just don't consider it production ready (at least not officially).

dionorgua commented:
Hi,

@bertbesser 

It was working pretty well on my machine in time it was created (but yes, it's very experimental). Good news is that it doesn't touch any data pack files at all. Basically It just adds new snapshot as copy of existing with a few files removed (plus optionally removes old snapshot). I don't think that it can corrupt repository (except new snapshots, that can be just deleted manually)

I think that 'rewrite + check + prune' scenario should be safe enough. I've already used it on my own 2.5TB repository without any problems. Removed most of garbage from old snapshots.

But I'm not sure that it'll work as is right now, I'll rebase it on top of latest git and try to address most of comments next week.

edmwagner commented:
Absolutely!  I’ve also used it to clean up 30gb of junk from early
backups.  I didn’t know how to filter back then.  My repo was 40 gb and now
rest consistently under 10 with about 2years of backups.  Really helped!!!
But that was also some time ago.

On Thu, Oct 21, 2021 at 3:26 PM Gaibhne ***@***.***> wrote:

> This looks like a fantastic feature I sorely miss in Restic. Is there
> anything I, as a user, can do to help with this PR ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/restic/restic/pull/2731#issuecomment-948933055>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACPBAZEXMQKIZCY5DGQ4YRTUIBSNNANCNFSM4M6YEN6A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

komic commented:
Hi @dionorgua, is there any chance you'd have an update regarding your work on this PR? I may or may have not oopsied a B2 backup again. 🤠

MichaelEischer committed:
rewrite: fix compilation

MichaelEischer committed:
rewrite: address most review comments

MichaelEischer committed:
rewrite: use treejsonbuilder

MichaelEischer committed:
rewrite: simplify dryrun

MichaelEischer committed:
rewrite: extract tree filtering

MichaelEischer committed:
rewrite: filter all snapshots if none are specified

MichaelEischer committed:
unify exclude patterns with backup command

MichaelEischer committed:
unify exclude pattern options

MichaelEischer committed:
rewrite: start repository uploader goroutines

MichaelEischer committed:
rewrite: use SelectByName like in the backup command

MichaelEischer committed:
rewrite: cleanup err handling and output

MichaelEischer committed:
rewrite: non-exclusive lock if snapshots are only added

MichaelEischer committed:
rewrite: add minimal test

MichaelEischer commented:
I've rebased the PR and made quite a few cleanups (see the commit messages for more details). The main tasks left are:
- [ ] Writing some documentation
- [ ] More tests. The current test coverage is just the absolute minimum
- [ ] Decide what to do with the `--inplace` flag, I'll probably rename it to `--forget`
- [ ] Performance optimizations.
- [ ] Finalize the CLI options

MichaelEischer commented on cmd/restic/cmd_rewrite.go:
```diff
@@ -0,0 +1,296 @@
+	f.StringArrayVarP(&rewriteOptions.Hosts, "host", "H", nil, "only consider snapshots for this `host`, when no snapshot ID is given (can be specified multiple times)")
+	f.Var(&rewriteOptions.Tags, "tag", "only consider snapshots which include this `taglist`, when no snapshot-ID is given")
+	f.StringArrayVar(&rewriteOptions.Paths, "path", nil, "only consider snapshots which include this (absolute) `path`, when no snapshot-ID is given")
+	f.BoolVarP(&rewriteOptions.Inplace, "inplace", "", false, "replace existing snapshots")
```
Adding all prune options would clutter the options list for the `rewrite` command. I think it's better to keep them separate.

NovacomExperts commented:
I'm so thrilled that this PR is getting some progress ! Many thanks to the team members for all the latest commits. I will do some test on my mostly "local" repos.

I support @MichaelEischer on the naming for `--forget` instead of `--inplace`. Of course, "are you sure" could follow before execution, just like prune. As discussed with @dionorgua before, a clear `--tag-old-snapshots-with <tag>` and `--tag-new-snapshots-with <tag>` would work great for managing 2 step forget process. 

Keep on the great work !

bertbesser commented:
> Hi,
> 
> @bertbesser
> 
> It was working pretty well on my machine in time it was created (but yes, it's very experimental). Good news is that it doesn't touch any data pack files at all. Basically It just adds new snapshot as copy of existing with a few files removed (plus optionally removes old snapshot). I don't think that it can corrupt repository (except new snapshots, that can be just deleted manually)
> 
> I think that 'rewrite + check + prune' scenario should be safe enough. I've already used it on my own 2.5TB repository without any problems. Removed most of garbage from old snapshots.
> 
> But I'm not sure that it'll work as is right now, I'll rebase it on top of latest git and try to address most of comments next week.

@NovacomExperts The `--tag-xxx-snapshots-with` would just be a welcome addition to this, correct? Not a fundamentally other approch if I get you right. Cheers

MichaelEischer committed:
rewrite: use unified snapshot filter options

MichaelEischer committed:
rewrite: rename --inplace to --forget

MichaelEischer committed:
walker: Add tests for FilterTree

MichaelEischer committed:
rewrite: Add tests for further ways to use the command

MichaelEischer committed:
rewrite: add documentation

MichaelEischer commented:
I've added enough tests to achieve over 80% test coverage for the new code. There's now also a new documentation section. I'll probably split some of the preparatory refactoring commits into separate PRs to reduce the amount of code a bit.

> As discussed with @dionorgua before, a clear `--tag-old-snapshots-with <tag>` and

`--tag-old-snapshots` is impossible to implement. We cannot change the tag of an existing snapshot without  creating a new one and removing the old snapshot.

Instead of using tags, using the printed snapshot id should also work.

MichaelEischer committed:
rewrite: Fail if a tree contains an unknown field

In principle, the JSON format of Tree objects is extensible without
requiring a format change. In order to not loose information just play
it safe and reject rewriting trees for which we could loose data.

MichaelEischer commented:
I've added a commit to let the `rewrite` command reject rewriting anything which would result in loosing data in a tree. This ensures that we can add new fields to the `Tree` struct without having to increase the repository version every time. Without the check, the `rewrite` command could just loose data.

The check is not completely optimal, but it should prevent rewriting for all problematic situations. If the need arises to handle this in a more targeted way, we can still adapt the implementation later on.

NovacomExperts commented:
@MichaelEischer This is such a great addition to Restic. Again, thank you a thousand times for making this possible. At a point I was almost ready to learn GO and give it a shot, but seeing you commits, and the level of expertise and care needed... Man this is waaaay out my league :)

`--tag-old-snapshots` was merely a brainstorming idea. I think that the --forget  flag will do the job perfectly in one time.

I'll have some free time to in the next few days to recompile with the latest commits. I'll test a few of my "large" repos.

rawtaz committed:
rewrite: Revert unrelated documentation change

rawtaz committed:
rewrite: Polish code and add missing messages

rawtaz committed:
rewrite: Polish documentation

NovacomExperts commented:
Hello !

I've been playing around with [72196e5](https://github.com/restic/restic/commit/72196e54c0e71d4d2ae26314173f57914b09ba28) and I had to use in production on a version 1 repo. It was a 3 host local repo composed of 830 GB and was correctly shrinked to 319 GB. Everything went very well and all checks are ok (including read-data). Restore tests were OK too. I have kept a backup of the original repo just in case but so far so good ! I wanted to exclude a specific folder from multiple snapshots **except the last 30 days** on a specific host. Here’s what I ended up doing :

`restic forget --dry-run --host myhost --keep-last 30`

Copied the list of snapshots listed a the end _without the brackets_

`restic rewrite --forget --exclude /home/test/foldertoexclude <pastesnapshots>`

Take a coffee, or two

`restic prune`

enjoy !

MichaelEischer committed:
rewrite: Fix check that an exclude pattern was passed

The old check did not consider files containing case insensitive
excludes. The check is now implemented as a function of the
excludePatternOptions struct to improve cohesion.

MichaelEischer committed:
Cleanup exclude pattern collection

MichaelEischer commented:
LGTM. I've made one final change before merging this PR: when just passing`--iexclude-file` the `rewrite` command would complain about missing excludes.

> Everything went very well and all checks are ok (including read-data).

Thanks for testing! Good to hear that the PR indeed worked as expected (it's been in a good shape for some time.

MichaelEischer reviewed (approved):
LGTM2

12nick12 commented:
Any idea when this will be in the prebuilt binaries?

rawtaz commented:
It will be in the next release, which is released when it's done.

aawsome commented:
@12nick12  You might want to have a look at https://beta.restic.net/

rawtaz commented:
FWIW, we're very close to a new release so stay tuned and you'll have it in your hands soon :)

rawtaz commented:
@12nick12 It's out now! See: https://restic.net/blog/2023-01-12/restic-0.15.0-released/

vvidovic commented:
> @12nick12 It's out now! See: https://restic.net/blog/2023-01-12/restic-0.15.0-released/

💚💚💚💚
